tests/handlers/test_sync.py

```diff
@@ -186,6 +187,139 @@ def test_unknown_room_version(self):
+        # - the homeserver generates a join event with prev_events that precede the ban
+        #   (so that it passes the "are you banned" test)
+        # - but the join event has a stream_ordering after that of the ban (so that it
+        #   comes down /sync after the ban).
+        #
+        # We do this by
+        # 1. Ban eve from the room.
+        # 2. Have eve join the room.
+        #    - mock the prev_events for the join event to precede the ban
+
+        # First, the ban.
+        eve = self.register_user("eve", "password")
+        eve_token = self.login(eve, "password")
+        self.get_success(
+            inject_event(
+                self.hs,
+                sender=moderator,
+                prev_event_ids=[moderator_pl_event_id],
+                type=EventTypes.Member,
+                state_key=eve,
+                room_id=room_id,
+                content={"membership": "ban"},
+            )
+        )
+
+        # We need a sync_token for eve after the ban.
+        eve_requester = create_requester(eve)
+        eve_sync_config = generate_sync_config(eve)
+        eve_sync_after_ban: SyncResult = self.get_success(
+            self.sync_handler.wait_for_sync_for_user(eve_requester, eve_sync_config)
+        )
+        since_token = eve_sync_after_ban.next_batch
+        # Sanity check this sync result. We should be banned from the room.
+        self.assertEqual(len(eve_sync_after_ban.joined), 0)
+        self.assertEqual(len(eve_sync_after_ban.archived), 1)
+        self.assertEqual(eve_sync_after_ban.archived[0].room_id, room_id)
+
+        mocked_get_prev_events = patch.object(
+            self.hs.get_datastore(),
+            "get_prev_events_for_room",
+            new_callable=MagicMock,
+            return_value=make_awaitable([moderator_pl_event_id]),
+        )
+        with mocked_get_prev_events:
+            self.helper.join(room_id, eve, tok=eve_token)
+
+        # Eve makes a second, incremental sync. The join event should appear, but not
+        # under the "joined" section.
+        eve_incremental_sync_after_join: SyncResult = self.get_success(
+            self.sync_handler.wait_for_sync_for_user(
+                eve_requester,
+                eve_sync_config,
+                since_token=since_token,
+            )
+        )
+        # The incremental sync should have nothing to show.
+        self.assertEqual(len(eve_incremental_sync_after_join.joined), 0)
+        self.assertEqual(len(eve_incremental_sync_after_join.archived), 0)
+
+        # If we did a third initial sync, we should _still_ see that eve is banned.
+        eve_initial_sync_after_join: SyncResult = self.get_success(
+            self.sync_handler.wait_for_sync_for_user(
+                eve_requester,
+                eve_sync_config,
+            )
+        )
+        self.assertEqual(len(eve_initial_sync_after_join.joined), 0)
+        self.assertEqual(len(eve_initial_sync_after_join.archived), 1)
```

Comment:
I'm confused at why this seems to flip in and out of archived, something here feels buggy, but maybe it is working as expected?

Reply:
There are three syncs:

1. `eve_sync_after_ban`, an  initial sync,
2. `eve_incremental_sync_after_join`, which follows (1), and
3. `eve_initial_sync_after_join`.

(Maybe the naming is too verbose here?)

(1 + 2) and (3) are meant to cover the same range of events in the room. I think (1+2) and (3) are consistent.

> maybe it is working as expected?

maybe; again, I'm not sure we know what "expected" actually is.

Reply:
> > I think (1+2) and (3) are consistent.
> 
> Consistent in that 2 shows no change since the state of the room has not changed according to eve?

Yes, exactly this.